packages/libsql/src/lib/Category.ts

```diff
@@ -0,0 +1,417 @@
+import { randomUUID } from "crypto"
+
+const keepMap = Equinox.Internal.keepMap
+
+type GatewaySyncResult = { type: "Written"; token: StreamToken } | { type: "ConflictUnknown" }
+
+type Decode<E> = (v: ITimelineEvent<Format>) => E | undefined
+
+export class LibSqlConnection {
+  constructor(
+    public read: LibSqlReader,
+    public write: LibSqlWriter,
+  ) {}
+
+  static create(writeClient: Client, readClient = writeClient) {
+    return new LibSqlConnection(new LibSqlReader(readClient), new LibSqlWriter(writeClient))
+  }
+}
+
+type ContextConfig = {
+  client: Client
+  readClient?: Client
+  batchSize: number
+  maxBatches?: number
+}
+
+export class LibSqlContext {
+  constructor(
+    private readonly conn: LibSqlConnection,
+    public readonly batchSize: number,
+    public readonly maxBatches?: number,
+  ) {}
+
+  tokenEmpty = Token.create(0n)
+
+  async loadBatched<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    let state = initial
+    let version = 0n
+    let batches = 0
+    let eventCount = 0
+    for await (const [lastVersion, events] of Read.loadForwardsFrom(
+      this.conn.read,
+      this.batchSize,
+      this.maxBatches,
+      streamName,
+      0n,
+    )) {
+      batches++
+      eventCount += events.length
+      state = fold(state, keepMap(events, decode))
+      version = lastVersion
+    }
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: eventCount,
+      [Tags.batches]: batches,
+      [Tags.read_version]: Number(version),
+    })
+    return [Token.create(version), state]
+  }
+
+  async loadLast<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    const [version, events] = await Read.loadLastEvent(this.conn.read, streamName)
+    trace.getActiveSpan()?.setAttributes({
+      [Tags.loaded_count]: 1,
+      [Tags.read_version]: Number(version),
+    })
+    return [Token.create(version), fold(initial, keepMap(events, decode))]
+  }
+
+  async loadState<Event, State>(
+    streamName: string,
+    decode: Decode<Event>,
+    fold: (state: State, events: Event[]) => State,
+    initial: State,
+  ): Promise<[StreamToken, State]> {
+    const snapshot = await this.conn.read.readSnapshot(streamName)
+    // An end-user might add RollingState to a category after the fact, we should handle this gracefully
+    if (snapshot == null) return this.loadBatched(streamName, decode, fold, initial)
+    const [etag, event] = snapshot
+    const decoded = decode(event)
+    // An end-user might add RollingState to a category after the fact, we should handle this gracefully
+    if (!decoded) return this.loadBatched(streamName, decode, fold, initial)
+    const version = BigInt(event.index)
+    const state = fold(initial, [decoded])
```

Comment:
should you pass null instead? I know I dont do that consistently and it could be very frustrating to hit in prod, but if a random evolve from snapshot impl is assuming that the incoming state is initial (or something else), that's a recipe for a very long bug hunt and/or loss of trust

Reply:
I know F# allows you to do unsafe stuff like that, but TS does not. If I pass a `null` the compiler will yell at me because it is not a valid `State` value